wsmaster/che-core-api-system/src/main/java/org/eclipse/che/api/system/server/ServiceTerminator.java

```diff
@@ -44,16 +54,110 @@
   void terminateAll() throws InterruptedException {
     for (ServiceTermination termination : terminations) {
       LOG.info("Shutting down '{}' service", termination.getServiceName());
-      eventService.publish(new StoppingSystemServiceEvent(termination.getServiceName()));
+      doTerminate(termination);
+    }
+  }
+
+  /**
+   * Suspends system services.
+   *
+   * @throws InterruptedException when suspending is interrupted
+   */
+  void suspendAll() throws InterruptedException {
+    for (ServiceTermination termination : terminations) {
+      LOG.info("Suspending down '{}' service", termination.getServiceName());
+      eventService.publish(new SuspendingSystemServiceEvent(termination.getServiceName()));
       try {
-        termination.terminate();
+        termination.suspend();
+      } catch (UnsupportedOperationException e) {
+        LOG.info(
+            "Suspending down '{}' service isn't supported, terminating it",
+            termination.getServiceName());
+        doTerminate(termination);
       } catch (InterruptedException x) {
         LOG.error(
-            "Interrupted while waiting for '{}' service to shutdown", termination.getServiceName());
+            "Interrupted while waiting for '{}' service to suspend", termination.getServiceName());
         throw x;
       }
-      LOG.info("Service '{}' is shut down", termination.getServiceName());
-      eventService.publish(new SystemServiceStoppedEvent(termination.getServiceName()));
+      LOG.info("Service '{}' is suspended", termination.getServiceName());
+      eventService.publish(new SystemServiceSuspendedEvent(termination.getServiceName()));
+    }
+  }
+
+  @VisibleForTesting
+  void doTerminate(ServiceTermination termination) throws InterruptedException {
+    eventService.publish(new StoppingSystemServiceEvent(termination.getServiceName()));
+    try {
+      termination.terminate();
+    } catch (InterruptedException x) {
+      LOG.error(
+          "Interrupted while waiting for '{}' service to shutdown", termination.getServiceName());
+      throw x;
+    }
+    LOG.info("Service '{}' is shut down", termination.getServiceName());
+    eventService.publish(new SystemServiceStoppedEvent(termination.getServiceName()));
+  }
+
+  private void checkNamesAndDependencies(Set<ServiceTermination> terminationSet) {
+    Set<String> uniqueNamesSet = new HashSet<>();
+    terminationSet.forEach(
+        t -> {
+          if (!uniqueNamesSet.add(t.getServiceName())) {
+            throw new RuntimeException(
+                String.format(
+                    "Duplicate termination found with service name %s", t.getServiceName()));
+          }
+        });
+
+    terminationSet.forEach(
+        t -> {
+          if (!uniqueNamesSet.containsAll(t.getDependencies())) {
+            throw new RuntimeException(
```

Comment:
I guess ConfigurationException is supposed to be used in cases like that.

Reply:
Not sure here... there is nothing to configure in this case...